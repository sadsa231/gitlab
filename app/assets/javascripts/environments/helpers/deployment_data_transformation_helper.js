import { getIdFromGraphQLId } from '~/graphql_shared/utils';

/**
 * This function transforms Commit object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue author object
 * @param {Object} Commit
 * @returns {Object}
 */
export const getAuthorFromCommit = (commit) => {
  if (commit.author) {
    return {
      username: commit.author.name,
      path: commit.author.webUrl,
      avatar_url: commit.author.avatarUrl,
    };
  }
  return {
    username: commit.authorName,
    path: `mailto:${commit.authorEmail}`,
    avatar_url: commit.authorGravatar,
  };
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const getCommitFromDeploymentNode = (deploymentNode) => {
  if (!deploymentNode.commit) {
    throw new Error("deploymentNode argument doesn't have 'commit' field", deploymentNode);
  }
  return {
    title: deploymentNode.commit.message,
    commitUrl: deploymentNode.commit.webUrl,
    shortSha: deploymentNode.commit.shortId,
    tag: deploymentNode.tag,
    commitRef: {
      name: deploymentNode.ref,
    },
    author: getAuthorFromCommit(deploymentNode.commit),
  };
};

export const convertJobToDeploymentAction = (job) => {
  return {
    name: job.name,
    playable: job.playable,
    scheduledAt: job.scheduledAt,
    playPath: `${job.webPath}/play`,
  };
};

export const getActionsFromDeploymentNode = (deploymentNode, lastDeploymentName) => {
  if (!deploymentNode || !lastDeploymentName) {
    return [];
  }

  return (
    deploymentNode.job?.deploymentPipeline?.jobs?.nodes
      ?.filter((deployment) => deployment.name !== lastDeploymentName)
      .map(convertJobToDeploymentAction) || []
  );
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/environments/environment_details/page.vue table
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const convertToDeploymentTableRow = (deploymentNode, environment) => {
  const { lastDeployment } = environment;
  const commit = getCommitFromDeploymentNode(deploymentNode);
  return {
    status: deploymentNode.status.toLowerCase(),
    id: deploymentNode.iid,
    triggerer: deploymentNode.triggerer,
    commit,
    job: deploymentNode.job && {
      webPath: deploymentNode.job.webPath,
      label: `${deploymentNode.job.name} (#${getIdFromGraphQLId(deploymentNode.job.id)})`,
    },
    created: deploymentNode.createdAt || '',
    deployed: deploymentNode.finishedAt || '',
    actions: getActionsFromDeploymentNode(deploymentNode, lastDeployment?.job?.name),
  };
};
