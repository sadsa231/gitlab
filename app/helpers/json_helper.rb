# frozen_string_literal: true

module JsonHelper
  # These two JSON helpers are short-form wrappers for the Gitlab::Json
  # class, which should be used in place of .to_json calls or calls to
  # the JSON class.
  def json_generate(*args)
    Gitlab::Json.generate(*args)
  end

  def json_parse(*args)
    Gitlab::Json.parse(*args)
  end
end
