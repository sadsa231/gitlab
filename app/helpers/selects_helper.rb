# frozen_string_literal: true

module SelectsHelper
  def project_select_tag(id, opts = {})
    opts[:class] = [*opts[:class], 'ajax-project-select'].join(' ')

    unless opts.delete(:scope) == :all
      if @group
        opts['data-group-id'] = @group.id
      end
    end

    with_feature_enabled_data_attribute =
      case opts.delete(:with_feature_enabled)
      when 'issues'         then 'data-with-issues-enabled'
      when 'merge_requests' then 'data-with-merge-requests-enabled'
      end

    opts[with_feature_enabled_data_attribute] = true

    hidden_field_tag(id, opts[:selected], opts)
  end

  def select2_tag(id, opts = {})
    klass_opts = [opts[:class]]
    klass_opts << 'multiselect' if opts[:multiple]

    opts[:class] = klass_opts.join(' ')
    value = opts[:selected] || ''

    hidden_field_tag(id, value, opts)
  end
end

SelectsHelper.prepend_mod_with('SelectsHelper')
