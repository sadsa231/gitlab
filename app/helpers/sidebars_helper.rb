# frozen_string_literal: true

module SidebarsHelper
  include Nav::NewDropdownHelper

  def sidebar_tracking_attributes_by_object(object)
    sidebar_attributes_for_object(object).fetch(:tracking_attrs, {})
  end

  def sidebar_qa_selector(object)
    sidebar_attributes_for_object(object).fetch(:sidebar_qa_selector, nil)
  end

  def scope_qa_menu_item(object)
    sidebar_attributes_for_object(object).fetch(:scope_qa_menu_item, nil)
  end

  def scope_avatar_classes(object)
    %w[avatar-container rect-avatar s32].tap do |klasses|
      klass = sidebar_attributes_for_object(object).fetch(:scope_avatar_class, nil)
      klasses << klass if klass
    end
  end

  def project_sidebar_context(project, user, current_ref, ref_type: nil)
    context_data = project_sidebar_context_data(project, user, current_ref, ref_type: ref_type)
    Sidebars::Projects::Context.new(**context_data)
  end

  def group_sidebar_context(group, user)
    context_data = group_sidebar_context_data(group, user)

    Sidebars::Groups::Context.new(**context_data)
  end

  def super_sidebar_context(user, group:, project:)
    {
      name: user.name,
      username: user.username,
      avatar_url: user.avatar_url,
      assigned_open_issues_count: user.assigned_open_issues_count,
      assigned_open_merge_requests_count: user.assigned_open_merge_requests_count,
      todos_pending_count: user.todos_pending_count,
      issues_dashboard_path: issues_dashboard_path(assignee_username: user.username),
      create_new_menu_groups: create_new_menu_groups(group: group, project: project),
      support_path: support_url,
      display_whats_new: display_whats_new?
    }
  end

  private

  def create_new_menu_groups(group:, project:)
    new_dropdown_sections = new_dropdown_view_model(group: group, project: project)[:menu_sections]
    show_headers = new_dropdown_sections.length > 1
    new_dropdown_sections.map do |section|
      {
        name: show_headers ? section[:title] : '',
        items: section[:menu_items].map do |item|
          {
            text: item[:title],
            href: item[:href]
          }
        end
      }
    end
  end

  def sidebar_attributes_for_object(object)
    case object
    when Project
      sidebar_project_attributes
    when Group
      sidebar_group_attributes
    when User
      sidebar_user_attributes
    else
      {}
    end
  end

  def sidebar_project_attributes
    {
      tracking_attrs: sidebar_project_tracking_attrs,
      sidebar_qa_selector: 'project_sidebar',
      scope_qa_menu_item: 'Project scope',
      scope_avatar_class: 'project_avatar'
    }
  end

  def sidebar_group_attributes
    {
      tracking_attrs: sidebar_group_tracking_attrs,
      sidebar_qa_selector: 'group_sidebar',
      scope_qa_menu_item: 'Group scope',
      scope_avatar_class: 'group_avatar'
    }
  end

  def sidebar_user_attributes
    {
      tracking_attrs: sidebar_user_profile_tracking_attrs
    }
  end

  def sidebar_project_tracking_attrs
    tracking_attrs('projects_side_navigation', 'render', 'projects_side_navigation')
  end

  def sidebar_group_tracking_attrs
    tracking_attrs('groups_side_navigation', 'render', 'groups_side_navigation')
  end

  def sidebar_user_profile_tracking_attrs
    tracking_attrs('user_side_navigation', 'render', 'user_side_navigation')
  end

  def project_sidebar_context_data(project, user, current_ref, ref_type: nil)
    {
      current_user: user,
      container: project,
      current_ref: current_ref,
      ref_type: ref_type,
      jira_issues_integration: project_jira_issues_integration?,
      can_view_pipeline_editor: can_view_pipeline_editor?(project),
      show_cluster_hint: show_gke_cluster_integration_callout?(project)
    }
  end

  def group_sidebar_context_data(group, user)
    {
      current_user: user,
      container: group
    }
  end
end

SidebarsHelper.prepend_mod_with('SidebarsHelper')
