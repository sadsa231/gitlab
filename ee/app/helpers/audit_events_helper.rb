# frozen_string_literal: true

module AuditEventsHelper
  FILTER_TOKEN_TYPES = {
    user: :user,
    group: :group,
    project: :project,
    member: :member
  }.freeze

  def admin_audit_event_tokens
    [
      { type: FILTER_TOKEN_TYPES[:user] },
      { type: FILTER_TOKEN_TYPES[:group] },
      { type: FILTER_TOKEN_TYPES[:project] }
    ].freeze
  end

  def group_audit_event_tokens(group_id)
    [{ type: FILTER_TOKEN_TYPES[:member], group_id: group_id }].freeze
  end

  def project_audit_event_tokens(project_path)
    [{ type: FILTER_TOKEN_TYPES[:member], project_path: project_path }].freeze
  end

  def export_url
    admin_audit_log_reports_url(format: :csv)
  end

  def view_only_own_project_events?(project)
    !can?(current_user, :admin_project, project) && !current_user.auditor?
  end

  def view_only_own_group_events?(group)
    !can?(current_user, :admin_group, group) && !current_user.auditor?
  end

  def filter_view_only_own_events_token_values(view_only)
    return [] unless view_only

    [{ type: FILTER_TOKEN_TYPES[:member], data: "@#{current_user.username}" }]
  end

  def show_streams_for_group?(group)
    return false if group.subgroup?

    can?(current_user, :admin_external_audit_events, group)
  end
end
