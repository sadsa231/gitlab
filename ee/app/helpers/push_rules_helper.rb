# frozen_string_literal: true

module PushRulesHelper
  def can_modify_group_push_rules?(current_user, group)
    can?(current_user, :change_push_rules, group)
  end

  def reject_unsigned_commits_description(push_rule)
    message = s_("ProjectSettings|Only signed commits can be pushed to this repository.")

    push_rule_update_description(message, push_rule, :reject_unsigned_commits)
  end

  def reject_non_dco_commits_description(push_rule)
    message = format(s_("ProjectSettings|Only commits that include a %{code_block_start}Signed-off-by:%{code_block_end} element can be pushed to this repository.").html_safe, code_block_start: '<code>'.html_safe, code_block_end: '</code>'.html_safe)

    push_rule_update_description(message, push_rule, :reject_non_dco_commits)
  end

  def commit_committer_check_description(push_rule)
    message = s_("ProjectSettings|Users can only push commits to this repository "\
                 "if the committer email is one of their own verified emails.")

    push_rule_update_description(message, push_rule, :commit_committer_check)
  end

  def commit_committer_name_check_description(push_rule)
    message = s_("ProjectSettings|Users can only push commits to this repository "\
      "if the commit author name is consistent with their GitLab account name.")

    push_rule_update_description(message, push_rule, :commit_committer_name_check)
  end

  private

  def push_rule_update_description(message, push_rule, rule)
    messages = [message]
    if push_rule.global?
      messages << s_("ProjectSettings|This setting will be applied to all projects unless overridden by an admin.")
    else
      enabled_globally = PushRule.global&.public_send(rule)
      enabled_in_project = push_rule.public_send(rule) # rubocop:disable GitlabSecurity/PublicSend

      if enabled_globally
        messages << if enabled_in_project
                      s_("ProjectSettings|This setting is applied on the server level and can be overridden by an admin.")
                    else
                      s_("ProjectSettings|This setting is applied on the server level but has been overridden for this project.")
                    end

        unless current_user.can_admin_all_resources?
          messages << s_("ProjectSettings|Contact an admin to change this setting.")
        end
      end
    end

    messages.join(' ').html_safe
  end
end
