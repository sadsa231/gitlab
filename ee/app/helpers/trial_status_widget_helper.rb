# frozen_string_literal: true

# NOTE: The patterns first introduced in this helper for doing trial-related
# callouts are mimicked by the PaidFeatureCalloutHelper. A third reuse of these
# patterns (especially as these experiments finish & become permanent parts of
# the codebase) could trigger the need to extract these patterns into a single,
# reusable, sharable helper.
module TrialStatusWidgetHelper
  def trial_status_popover_data_attrs(group, trial_status, ultimate_plan_id)
    hand_raise_attrs =
      experiment(
        :group_contact_sales, namespace: group.root_ancestor, user: current_user, sticky_to: current_user
      ) do |e|
        e.control { {} }
        e.candidate { hand_raise_props(group, glm_content: 'trial-status-show-group') }
      end.run

    base_attrs = trial_status_common_data_attrs(group).merge(hand_raise_attrs)

    base_attrs.merge(
      days_remaining: trial_status.days_remaining, # for experiment tracking
      group_name: group.name,
      purchase_href: ultimate_subscription_path_for_group(group, ultimate_plan_id),
      target_id: base_attrs[:container_id],
      trial_end_date: trial_status.ends_on
    )
  end

  def trial_status_widget_data_attrs(group, trial_status)
    trial_status_common_data_attrs(group).merge(
      trial_days_used: trial_status.days_used,
      trial_duration: trial_status.duration,
      nav_icon_image_path: image_path('illustrations/golden_tanuki.svg'),
      percentage_complete: trial_status.percentage_complete
    )
  end

  private

  def trial_status_common_data_attrs(group)
    {
      container_id: 'trial-status-sidebar-widget',
      plan_name: group.gitlab_subscription.plan_title,
      plans_href: group_billings_path(group)
    }
  end

  def ultimate_subscription_path_for_group(group, ultimate_plan_id)
    new_subscriptions_path(namespace_id: group.id, plan_id: ultimate_plan_id)
  end
end

TrialStatusWidgetHelper.prepend_mod
