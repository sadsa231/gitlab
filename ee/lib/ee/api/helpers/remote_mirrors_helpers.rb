# frozen_string_literal: true

module EE
  module API
    module Helpers
      module RemoteMirrorsHelpers
        extend ActiveSupport::Concern
        extend ::Gitlab::Utils::Override

        prepended do
          params :mirror_branches_setting_ee do
            optional :mirror_branch_regex, type: String, desc: 'Determines if only matched branches are mirrored'
            mutually_exclusive :only_protected_branches, :mirror_branch_regex
          end
        end

        override :verify_mirror_branches_setting
        def verify_mirror_branches_setting(attrs, project)
          attrs[:mirror_branch_regex] = nil unless ::Feature.enabled?(:mirror_only_branches_match_regex, project)

          if attrs[:only_protected_branches]
            attrs[:mirror_branch_regex] = nil
          elsif attrs[:mirror_branch_regex].present?
            attrs[:only_protected_branches] = false
          end
        end
      end
    end
  end
end
