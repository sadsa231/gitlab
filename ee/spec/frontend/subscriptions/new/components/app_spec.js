import { nextTick } from 'vue';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import Component from 'ee/subscriptions/new/components/app.vue';
import OrderSummary from 'ee/subscriptions/new/components/order_summary.vue';
import Checkout from 'ee/subscriptions/new/components/checkout.vue';
import Modal from 'ee/subscriptions/new/components/modal.vue';
import { stubExperiments } from 'helpers/experimentation_helper';
import GitlabExperiment from '~/experimentation/components/gitlab_experiment.vue';

describe('App component', () => {
  let wrapper;

  const createComponent = () => {
    return shallowMountExtended(Component, {
      stubs: {
        Modal,
        GitlabExperiment,
        StepOrderApp: {
          template: `
            <div>
                <slot name="checkout"></slot>
                <slot name="order-summary"></slot>
            </div>
            `,
        },
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe('cart_abandonment_modal experiment', () => {
    describe('control', () => {
      beforeEach(() => {
        stubExperiments({ cart_abandonment_modal: 'control' });
        wrapper = createComponent();
      });

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders the modal', () => {
        expect(wrapper.findComponent(Modal).exists()).toBe(false);
      });
    });

    describe('candidate', () => {
      beforeEach(() => {
        stubExperiments({ cart_abandonment_modal: 'candidate' });
        wrapper = createComponent();
      });

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders the modal', () => {
        expect(wrapper.findComponent(Modal).exists()).toBe(true);
      });
    });
  });

  describe('step order app', () => {
    beforeEach(() => {
      wrapper = createComponent();
    });

    it('renders checkout', () => {
      expect(wrapper.findComponent(Checkout).exists()).toBe(true);
    });

    it('renders order summary', () => {
      expect(wrapper.findComponent(OrderSummary).exists()).toBe(true);
    });
  });

  describe('confirm order CTA', () => {
    it(`should show confirm order CTA`, async () => {
      wrapper = createComponent();
      await nextTick();

      expect(wrapper.findByTestId('confirm-order-desktop').classes()).toEqual([
        'gl-display-none',
        'gl-lg-display-block!',
      ]);
      expect(wrapper.findByTestId('confirm-order-mobile').classes()).toEqual([
        'gl-display-block',
        'gl-lg-display-none!',
      ]);
    });
  });
});
