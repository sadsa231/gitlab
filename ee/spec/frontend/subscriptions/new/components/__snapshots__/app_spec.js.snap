// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component cart_abandonment_modal experiment candidate matches the snapshot 1`] = `
<div>
  <gl-modal-stub
    arialabel=""
    dismisslabel="Close"
    modalclass=""
    modalid="subscription-modal"
    size="sm"
    titletag="h4"
  >
    

  We understand. Maybe you have some questions for our sales team, or maybe you'd like to try some of the paid features first. What would you like to do?

  
  </gl-modal-stub>
   
  <div>
    <checkout-stub />
     
    <confirm-order-stub
      class="gl-display-none gl-lg-display-block!"
      data-testid="confirm-order-desktop"
    />
     
    <order-summary-stub />
     
    <confirm-order-stub
      class="gl-display-block gl-lg-display-none!"
      data-testid="confirm-order-mobile"
    />
  </div>
</div>
`;

exports[`App component cart_abandonment_modal experiment control matches the snapshot 1`] = `
<div>
  <!---->
   
  <div>
    <checkout-stub />
     
    <confirm-order-stub
      class="gl-display-none gl-lg-display-block!"
      data-testid="confirm-order-desktop"
    />
     
    <order-summary-stub />
     
    <confirm-order-stub
      class="gl-display-block gl-lg-display-none!"
      data-testid="confirm-order-mobile"
    />
  </div>
</div>
`;
